Theme base 2

pngwn opened the pull request:
Not quite done but the base is pretty much sorted apart from for responsiveness + a few focus styles. I'm just applying the new system to everything.

Forms are pretty much done (with one or two wrinkles regarding focus). You can take a look at the kitchen sink to see if there are any inconsistencies but I'd hold off on spending _too_ much time on things just yet.

Will add more detail soon. Also need to move a few things around.

PR is huge and will get bigger, wouldn't bother trying to review just yet.

Please include: 
* relevant motivation
* a summary of the change 
* which issue is fixed. 
* any additional dependencies that are required for this change.

Closes: # (issue)

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes


# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

abidlabs commented:
Very cool @pngwn, started to test things out and sharing any discrepancies between `main` and this branch here:

* `blocks_layout`: there is some gray space that's appeared under the "Count" and "Radio" boxes

<img width="988" alt="image" src="https://user-images.githubusercontent.com/1778297/208331733-aabe1526-fa29-48d5-a003-7616c7b2131e.png">


* `kitchen_sink_random`: not sure why but it seems that I can't run this demo in this branch in dev mode

abidlabs commented:
Demos in which I could not find any difference before and after:
* `altair_plot`
* `audio_debugger`
* `blocks_essay`

abidlabs commented:
The typeable number box in the sliders have lost their border-radius, e.g. see `interface_random_slider`:

Before:

<img width="186" alt="image" src="https://user-images.githubusercontent.com/1778297/208832175-76b74cbd-f93f-4975-a14a-fcf1ffbf73a9.png">


After:

<img width="157" alt="image" src="https://user-images.githubusercontent.com/1778297/208832124-cad8626b-d9f7-4349-91e5-9571076c6d03.png">

freddyaboulton commented:
@pngwn When running `native_plots` demo locally, I see some differences:

1) Drop down borders are gone
2) The container is gone on the plot

### This branch
![image](https://user-images.githubusercontent.com/41651716/208983241-a328af4e-cc72-48b1-9bda-28fe3eb7a2f1.png)

### Main
![image](https://user-images.githubusercontent.com/41651716/208983274-e475865f-c667-4914-862e-641a6a375046.png)


Let me know if I've made a mistake in testing/building.

What I did:
1) Clone this branch
2) `pnpm i` in `ui/`
3) `pnpm build`
4) Launch demo. Comparing to version 3.15.0 installed on a separate environment

abidlabs commented:
Uhh okay so this is weird. But I was using the deployed PR Spaces to check, because I had noticed that for the `blocks_layout` demo, the deployed PR Space correctly reflected how the demo looked locally. 

But there's a big difference between the deployed PR Space depending on whether you open up the Space vs. the iframe URL directly. 

Space: https://huggingface.co/spaces/gradio-pr-deploys/pr-2840-all-demos: 

<img width="998" alt="image" src="https://user-images.githubusercontent.com/1778297/209003253-88f7e067-c974-49c5-911c-129ae0158aad.png">

iframe URL: https://gradio-pr-deploys-pr-2840-all-demos.hf.space/:

<img width="990" alt="image" src="https://user-images.githubusercontent.com/1778297/209003382-c5d373d7-9048-4dcb-9264-f57622d477db.png">

I would test locally, but I'm not able to run many of these demos locally (I use `scripts/run_frontend.sh` but the frontend just hangs). There might be many issues that I've missed sorry

pngwn commented:
In order to test:

```bash
cd ui
pnpm i
pnpm dev
```

In anotehr terminal:

```bash
cd demo/<some-demo>
gradio run.py
# or
python run.py
```

Open `localhost:3000` in a browser.

freddyaboulton commented on ui/packages/app/src/components/Accordion/Accordion.svelte:
```diff
@@ -14,14 +14,39 @@
+<style>
+	.container {
+		display: flex;
+		flex-direction: column;
```
There is a weird grey bar under the `count` and `radio` components of `blocks_layout`

### This branch
![image](https://user-images.githubusercontent.com/41651716/209225129-e8178bf4-a164-4cdf-8319-f7443b9968ec.png)

### 3.15.0
![image](https://user-images.githubusercontent.com/41651716/209225169-def65c2f-5a0e-4a4c-81c0-65d729411936.png)

freddyaboulton commented on ui/packages/app/src/components/Accordion/Accordion.svelte:
```diff
@@ -14,14 +14,39 @@
+		display: flex;
+		flex-direction: column;
+		gap: var(--size-3);
+		padding: var(--size-3);
```
Typeable value of slider in `blocks_update` is no longer centered

### This branch
![image](https://user-images.githubusercontent.com/41651716/209226753-317de2f3-34dd-4d8e-9fa5-ef006d16b2ee.png)


### 3.15.0
![image](https://user-images.githubusercontent.com/41651716/209226709-910a2f79-afb8-46c2-808b-010d4687a375.png)

freddyaboulton commented on ui/packages/app/src/components/Accordion/Accordion.svelte:
```diff
@@ -14,14 +14,39 @@
+		padding: var(--size-3);
+		border-width: 1px;
+		transition: 150ms;
+		border-radius: var(--radius-lg);
```
The default height of the chatbot in `chatbot_demo` is smaller

### 3.15.0
![image](https://user-images.githubusercontent.com/41651716/209230233-6cab6da7-a720-4311-aaf2-3b8d655bcae0.png)

### This branch
![image](https://user-images.githubusercontent.com/41651716/209230269-146c6d01-aa63-4069-ae6b-0485c052d521.png)

freddyaboulton commented on ui/packages/app/src/components/Accordion/Accordion.svelte:
```diff
@@ -14,14 +14,39 @@
+		border-radius: var(--radius-lg);
+	}
+
+	.label-wrap {
```
In Model3D the `x` in the input model3d component is smaller

### This branch
![image](https://user-images.githubusercontent.com/41651716/209231451-111f4662-a986-40a7-83cb-060ad1f0db30.png)

### 3.15.0
![image](https://user-images.githubusercontent.com/41651716/209231417-fa4b75e4-eba6-4c7f-b5b6-e9dc3ef30a67.png)


Also the file output component doesn't have a scrollbar anymore

### This branch
![image](https://user-images.githubusercontent.com/41651716/209231532-93c21768-8d9c-4c74-9d66-9e75cd239948.png)

### 3.15.0
![image](https://user-images.githubusercontent.com/41651716/209231572-77694eb3-1363-4f75-8368-7e6581d48422.png)

freddyaboulton reviewed (commented):
This is awesome @pngwn ! I ran about 20 demos and these are the only differences I could find. Most have no differences 👏

aliabid94 commented:
Been thinking about the Python API for all this, here's what I'm thinking:

1. Picking a predefined theme

```python
import gradio as gr

with gr.Blocks(theme=gr.themes.Monochrome()) as demo:
  pass
```

Pretty straightforward, just pick a theme from `gr.themes`

2. Modifying themes to create new themes

```python
import gradio as gr

theme2 = gr.themes.Default()
theme2.color.text.body = "red"
theme2.color.text.label = "brown"
theme2.color.background.primary = "pink"

with gr.Blocks(theme=theme2) as demo:
  pass
```
This approach allows users to explore available theme options with type-hinting, something that wouldn't be possible if they just used the css="" kwarg. They can also create reusable theme objects.
We'll need to keep the arguments of a theme object in sync with the css variables in the generated  theme.css. I'm thinking that `generate_theme.cjs` can create an additional JSON that the python can use.

Let me know what you guys think about this API, should be pretty straightforward to implement.

freddyaboulton commented:
@aliabid94 I like the theme kwarg. 

> gr.Blocks(theme=gr.themes.Monochrome())

For editing themes,

Instead of `theme2.color.text.body = "red"`, wouldn't `theme2.text.body.color = "red"` make more sense?

I think it should be easy to support community themes with this appoach. BTW, did you look at this community theme? https://github.com/wdcqc/doki-theme-gradio

freddyaboulton commented:
@pngwn I couldn't run any of the plot demos in dev mode

The demo on port 3000 hangs and I get the following in the console:

```
Uncaught (in promise) SyntaxError: The requested module '/@fs/Users/freddy/sources/gradio/ui/node_modules/.pnpm/clone@2.1.2/node_modules/clone/clone.js?v=3ce00a2c' does not provide an export named 'default' (at util.ts:1:9)
```

pngwn commented:
@freddyaboulton The addition of the altair plots broke dev mode (when using `PLOT`),I should have caught this when the PR went in but I only tested the build. In future we should make sure we don't break dev mode when we merge PRs if possible.

For now `Plot` just won't work in dev mode (I don't have a fix right now, it will require a bit more work), i have fixed the build now though, so I'll push that with the rest of the changes soon.

pngwn commented:
Okay, I think this work is complete for the most part. I have addressed all comments with some caveats.

Broadly speaking the UI should be identical both in style and function. There are a few differences.



**General**

- border-colors are now very consistent as they are pulled form the theme definition, we have a few and they are used everywhere. Previously there were inconsistencies that have no been addressed.
- Spacing is very slightly changed. Generally everything is the same but there have been some adjustment as our line height scales are slightly different which will sometimes make content taller/ shorter. The discrepancy that @freddyaboulton found has been fixed but the UI isn't identical to the pixel.
- Min-heights for components are now very consistent. This means that things should look a bit tidier and more consistent, previously every min-height was hardcode, this has been abstracted to some degree (where it makes sense). This is mostly relevant for `empty` states (with the icon). We now have an `empty` component with two size variants that can be passed an icon to display.
- The 'click or drag to upload' text is all very consistent now, this was hardcoded previously (and was styled inconsistently) now it is a single component for everything.
- We had a million different button implementations. Where it made sesne i used the `Button` component everywhere for consistency.
- Some gradients were slightly tweaked but are mostly the same.

**Tabs**

Tabs now scroll horizontally, I still don't love them but they are more functional when there are many:

Previously:

<img width="644" alt="Screenshot 2023-01-03 at 14 41 07" src="https://user-images.githubusercontent.com/12937446/210379307-577f6569-60b8-4eb6-ba89-c70e036111cc.png">

Now:

https://user-images.githubusercontent.com/12937446/210379481-c66a7435-9ad9-4fb8-a45c-b84521d14997.mov

**Error toast**

The previous error toast had way too many colours going on for me to fit into a sensible theme, I changed it. I can revert if we really need to but the colours weren't doing much anyway, they were too dark to really see as 'red'. 

Previously:

<img width="415" alt="Screenshot 2023-01-03 at 14 47 16" src="https://user-images.githubusercontent.com/12937446/210380412-6b91a2ca-fc1a-4a27-b648-4462f82af590.png">

Now:
<img width="514" alt="Screenshot 2023-01-03 at 14 46 40" src="https://user-images.githubusercontent.com/12937446/210380449-65775321-8547-458c-9475-8b38a909b255.png">

**Table row stripes**

The code for the stripes was there but there were no stripes, dno why. Tables now have alternative background colors for easier scanning (like the Dataset table):

Previously: 
<img width="654" alt="Screenshot 2023-01-03 at 14 55 41" src="https://user-images.githubusercontent.com/12937446/210382620-f439fba7-d823-4ad5-bbf3-7be4b2f148ff.png">

Now: 

<img width="654" alt="Screenshot 2023-01-03 at 14 55 33" src="https://user-images.githubusercontent.com/12937446/210382690-9293bf16-9e28-4503-9bca-5e1175c727f3.png">


**DataSet Dataframe**

This was all weird, fixed it:

Previously: 
https://user-images.githubusercontent.com/12937446/210382877-0996ffff-3149-4ecf-9664-d38df645c2b8.mov

Now: 

https://user-images.githubusercontent.com/12937446/210382913-405c6b4a-7cd2-464b-995e-3075231eae29.mov

**Typography** (mainly for markdown/ html)

I added some new styles for prose to keep things looking tidy (since the tw prose thing is gone now) and I also added a new font scale (for font sizes). The scale is smaller on mobile to keep things tighter. STill needs work but looks solid.

I expanded the current markdown demo to tests this so you are better running that. `deo/markdown_component`. Some of the styles pull from the theme (link link colors) but we can probably improve this over time.

**Other stuff**

Can't remember just leaving this as a placeholder.

**None style related**

I did a lot of refactoring, other than just the CSS. I pulled lots of things out into components to keep them consistent, I broke some larger components into multiple components (API docs is the most significant one). I tested pretty thorughly so i think everything works fine but let me know if there are any functional rather than style failures i should look at.

**Kind of breaky stuff**

Lot of the CSS classes have gone, if people were using the with the css kwarg their UIs might kind break. This is expected because they should be using the `elem_id` thing and descendant selectors to do any custom theming but I can almost guarantee that people will be using whatever css selectors we had on the DOM. There isn't anything we need to do here, just be aware of it in support requests. We have always said that we do not and never will version the DOM, so this is not a breaking change.

Will follow up with details about the actual theming system.

pngwn commented:
The new theme system and the approach to styling is relatively simple and requires no heavy tooling.

From this PR forwards we will no longer be using tailwind, just plain CSS and instead of classnames to apply standard values we will use CSS variables. These variables will also handle switching values for dark/ light mode.

The actual theme file needs a little cleanup but there are two categories of css variable, foundational tokens and theme values. Foundational tokens can be overwritten but are essentially always the same, this is for things like `size 1` size 2`, `font-mono` `font-sans`, etc . Theme tokens will have different values for light and dark mode (not always but they can do), this is for things like `primary border color` body font color` etc. Typically theme tokens will use foundational tokens but this isn't a hard and fast rule.

I need to figure out how to document the tokens, will think about that over the next day or so.

pngwn commented on ui/packages/app/src/api_docs/ApiBanner.svelte:
```diff
@@ -0,0 +1,65 @@
+<script lang="ts">
```
hmmmm

pngwn replied:
fixed. I did have to tweak the inputs a little, they are no consistent with out other inputs and have focus styles. Feedback welcome.

freddyaboulton commented on ui/packages/app/src/components/Dataset/Dataset.svelte:
```diff
@@ -61,36 +71,30 @@
-
 				<button
-					class="group rounded-lg"
+					class="button"
```
I think this removes the border around the example when there are only examples for one component?

Is that intentional?

### This PR
![image](https://user-images.githubusercontent.com/41651716/210423094-ac89bf59-ea0b-42c0-827a-85d2f2828d0b.png)

### Main
![image](https://user-images.githubusercontent.com/41651716/210423729-a89a734a-771d-480f-913e-18671eecaf29.png)

pngwn replied:
Ah, no this is a bug. I did see this and forgot to fix. Good catch.

freddyaboulton commented on ui/packages/app/src/api_docs/ApiDocs.svelte:
```diff
@@ -0,0 +1,176 @@
+<script lang="ts">
```
I think there is more whitespace now between the bottom of the demo and the `View API` button? This is with the model3D demo

### This PR
![view_api_whitespace_pr](https://user-images.githubusercontent.com/41651716/210424553-45cfa9e5-de17-4cc4-b1af-832ef1374273.gif)

### Main
![view_api_whitespace_main](https://user-images.githubusercontent.com/41651716/210424582-c521ddff-3887-475a-9075-8298b262910b.gif)

freddyaboulton commented on ui/packages/tabs/src/Tabs.svelte:
```diff
@@ -43,26 +43,116 @@
 	}
 
 	$: selected !== null && change_tab(selected);
+	let h = 0;
```
Can we add a scroll bar to the tabs? I found it difficult to scroll through all the tabs in the deployed demo manually with the current behavior.

pngwn replied:
Adding a scroll bar breaks the tab design because of where it appears. I can revert the change tho until we have a better design, if that is preferable.

freddyaboulton replied:
I vote we revert this change until we have a better design. Happy to keep if others feel strongly about it though!

pngwn replied:
and done!

freddyaboulton replied:
Thank you 🙌

freddyaboulton reviewed (commented):
@pngwn Thanks for addressing all of the comments and your hard work on this PR!! 

Noticed a few small issues, particularly about dark mode.  Looking forward to working on #2780 as that seems related to this PR.

Regarding what to add to the changelog, I think what you wrote [here](https://github.com/gradio-app/gradio/pull/2840#issuecomment-1369874339) would be a great inaugural blog post or newsletter article. Then we can add an entry to the changelog linking to that article. What do you think?

pngwn commented on ui/packages/app/src/components/StatusTracker/StatusTracker.svelte:
```diff
@@ -262,74 +261,249 @@
-						<button
-							on:click={close_message}
-							class="ml-auto text-gray-900 text-2xl pr-1">×</button
+					<button on:click={close_message} class="toast-close">
```
Which fading red?

freddyaboulton replied:
Sorry for the awful wording on my part 😂 

The old error pop-up had a fading red color in the header. Just wondering if we could keep the fading with the new design. Not blocking though. I like what you have now.

![image](https://user-images.githubusercontent.com/41651716/210437406-dc517c0b-8c7c-4c2d-ba3f-db7a8dacec22.png)

pngwn replied:
Added back for light mode, dark mode looked rly weird, so i've kept that plain for now.

pngwn commented on ui/packages/app/src/components/Dataset/ExampleComponents/Dataframe.svelte:
```diff
@@ -44,38 +49,100 @@
 {/if}
+
+<style>
+	.gallery {
```
I'm not seeing this with `transpose_matrix`, do you have repro steps? What browser/OS?

freddyaboulton replied:
I'm on Chrome with m1 mac.

Repro steps:

- Check out your branch
- `pnpm dev` from `ui`
- `python demo/matrix_transpose/run.py`
- Go to `localhost:3000`

It's only noticeable on dark mode though. Maybe that's related

pngwn replied:
I'm still not seeing this, regardless of what actions I take or screen size i view it with. Also tried light + dark mode.

<img width="662" alt="Screenshot 2023-01-04 at 20 44 07" src="https://user-images.githubusercontent.com/12937446/210646446-7d4f43ac-6b4f-4233-868b-87494b441e4c.png">

If you are still seeing this maybe we can jump on a call to debug it.

pngwn replied:
We did a call. Weird bug but fixed now.

pngwn commented on ui/packages/app/src/components/StatusTracker/StatusTracker.svelte:
```diff
@@ -223,8 +221,8 @@
 		{#if last_progress_level != null}
-			<div class="z-20 w-full flex items-center flex-col gap-1">
-				<div class="m-2 mx-auto font-mono text-xs dark:text-gray-100">
+			<div class="progress-level">
```
i really don't know what this fake diffusion is supposed to be doing, and i can' work it out from the old version. Could you provide more details please.

freddyaboulton replied:
Sorry I thought I uploaded a gif of what it looks like on main but looks like gh messes up rendering it. Here's what it looks like:

![fake_diffusion_main](https://user-images.githubusercontent.com/41651716/210464087-605f0a34-e4b4-4009-b551-c6928278b34a.gif)

It's supposed to show a random image every second and then the final image at the end. Throughout the iteration, the border should pulse orange.

The one thing that's different in this PR is that the whole image fades to white.

pngwn replied:
I get it now. Fixed.

pngwn commented:
leave me alone bot

pngwn commented:
Added a new demo in `demo/dataset` that shows all of the dataset types used in gallery mode (not table). The `kitchen_sink` demo shows tables mode. Everything seems to be working now and all comments addressed. Will resolve the new conflicts as soon as i can tomorrow.

aliabid94 commented:
I want to push back against the concept of component-specific theme values in themes (and the "foundation" vs "theme" config split). I think we should expand the "foundation" set of rules that are component-agnostic, and have all components internally refer to the this set of css variables, removing all component-specific rules from the theme. I believe this will vastly improve maintainability and usage of themes, because otherwise:

1. There are way too many theme values, making it difficult to create new themes, as each theme will have to set values for every component
2. Every time we add a new component, we will have to update all the existing themes
3. Component styling is no longer encapsulated within the folder for that component, (given that creating a new component requires updating all themes)

If there is further internal styling of components to be done, we can consider exposing them via the `style` function, but it should not be part of the theme.

I am happy to implement this refactoring over the next week as part of #2931 if you don't have issues with it @pngwn

pngwn commented:
Yeah I think most of those can go and the ones that do need to exists (I think some do but only a few) then they should pull from the foundational tokens so that theme values are applied everywhere even when the component specific ones aren't set. 

Places where dark and light mode variables are used and aren't consistent with other dark/ light mode values are an example of where they are needed for components.

pngwn commented:
I'll take care of this over the next day or two. I haven't had a chance to review the python API yet but I have a few queries and concerns that I'll post as soon as I can.

aliabid94 commented:
> Regarding what to add to the changelog, I think what you wrote https://github.com/gradio-app/gradio/pull/2840#issuecomment-1369874339 would be a great inaugural blog post or newsletter article. Then we can add an entry to the changelog linking to that article. What do you think?

For now, just added a simple entry to changelog since this doesn't enable theming yet.

freddyaboulton commented on gradio/components.py:
```diff
@@ -2838,6 +2838,9 @@ def style(
             **kwargs,
         )
 
+    def as_example(self, input_data: str | None) -> str:
```
Nice 👍

freddyaboulton reviewed (approved):
I went through a ton of demos this morning and they all match their appearance on main!

Thanks for all the hard work on this @pngwn !

aliabid94 commented:
Thanks for all the work @pngwn and thanks for testing as well @freddyaboulton! Things are looking good to me, wrapping up demo tests, will merge soon after.

aliabid94 commented:
just realized I missed one thing (don't think it was mentioned) - single component examples show as a table always, e.g instead of this:

![Screen Shot 2023-01-17 at 1 39 23 PM](https://user-images.githubusercontent.com/7870876/213018035-aa3b4628-bb74-48cc-85c7-317b6c1cfd7c.png)

it's like this

![Screen Shot 2023-01-17 at 1 39 35 PM](https://user-images.githubusercontent.com/7870876/213018081-bce9e2f5-83b0-45da-b251-9cfea8633643.png)

I do think the first was cleaner and more efficient real estate. I can restore this?

abidlabs commented:
Agreed 100%

superag commented:
> 

I had the exact some problem as two exactly same piece of code, when I run it alone, it shows in div, but when i put it inside automatic1111, it just keeps showing as table. what is trigger for that? searched for a long time and this page was what i find mentioning the same problem.
